{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d5e5e22",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "import os\n",
    "import shutil\n",
    "from glob import glob\n",
    "import zipfile\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "from shapely import wkb\n",
    "from shapely.geometry import shape\n",
    "import geopandas as gpd\n",
    "from shapely.geometry import Point\n",
    "\n",
    "def create_dir(path):\n",
    "    if not os.path.exists(path):\n",
    "        os.makedirs(path)\n",
    "QV_THRESHOLD = 20 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a87c7710",
   "metadata": {},
   "outputs": [],
   "source": [
    "# NCBI ÎßàÏª§ gene list\n",
    "marker_genes = {\n",
    "    \"epithelial\": [\n",
    "        # Epithelial Í∏∞Î≥∏ ÎßàÏª§\n",
    "        \"EPCAM\", \"KRT8\", \"KRT18\", \"KRT19\", \"KRT7\",\n",
    "        # Luminal ÎßàÏª§\n",
    "        \"ERBB2\", \"ESR1\", \"PGR\", \"GATA3\", \"FOXA1\",\n",
    "        # Proliferation\n",
    "        \"MKI67\", \"PCNA\",\n",
    "        # Junction\n",
    "        \"CDH1\", \"CLDN4\", \"CLDN3\", \"CLDN7\",\n",
    "        # Ï∂îÍ∞Ä ÏÉÅÌîº ÎßàÏª§\n",
    "        \"KRT5\", \"KRT14\", \"TP63\"\n",
    "    ],\n",
    "\n",
    "    \"Basal/Myoepithelial\": [\n",
    "        # Basal specific\n",
    "        \"KRT5\", \"KRT14\", \"KRT17\", \"TP63\", \"KRT15\",\n",
    "        # Myoepithelial specific\n",
    "        \"ACTA2\", \"MYH11\", \"MYLK\", \"TAGLN\", \"CNN1\", \"MYL9\",\n",
    "        \"OXTR\", \"CD10\", \"SMA\"\n",
    "    ],\n",
    "\n",
    "    \"Smooth muscle\": [\n",
    "        # Smooth muscle specific (vascular)\n",
    "        \"ACTA2\", \"MYH11\", \"MYLK\", \"TAGLN\", \"CNN1\", \"MYL9\",\n",
    "        \"ACTG2\", \"DES\", \"SMTN\", \"LMOD1\"\n",
    "    ],\n",
    "\n",
    "    \"Fibroblast\": [\n",
    "        # CAF (Cancer-Associated Fibroblast) ÎßàÏª§\n",
    "        \"PDGFRA\", \"PDGFRB\", \"FAP\", \"S100A4\", \"ACTA2\",\n",
    "        # ECM Í¥ÄÎ†®\n",
    "        \"COL1A1\", \"COL1A2\", \"COL3A1\", \"COL5A1\", \"COL6A1\",\n",
    "        \"FN1\", \"VIM\", \"DCN\", \"LUM\",\n",
    "        # Fibroblast specific\n",
    "        \"DPT\", \"SFRP1\", \"FBLN1\", \"SFRP4\", \"POSTN\",\n",
    "        \"THY1\", \"CDH2\", \"PDPN\"\n",
    "    ],\n",
    "\n",
    "    \"Endothelial\": [\n",
    "        # Pan-endothelial\n",
    "        \"PECAM1\", \"CD34\", \"CDH5\", \"VWF\", \"VCAM1\",\n",
    "        # Vascular\n",
    "        \"KDR\", \"FLT1\", \"CD93\", \"EGFL7\",\n",
    "        # Capillary/Lymphatic\n",
    "        \"CLEC14A\", \"MMRN2\", \"ESM1\", \"PLVAP\",\n",
    "        \"ACKR1\", \"SELE\", \"TEK\"\n",
    "    ],\n",
    "\n",
    "    \"Lymphocyte\": [\n",
    "        # T cell Î≤îÏö©\n",
    "        \"CD3D\", \"CD3E\", \"CD3G\", \"TRAC\", \"TRBC1\", \"TRBC2\",\n",
    "        # T cell subtype\n",
    "        \"CD4\", \"CD8A\", \"CD8B\",\n",
    "        # Activation/Function\n",
    "        \"GZMA\", \"GZMB\", \"GZMK\", \"NKG7\", \"CCL5\", \"PRF1\",\n",
    "        \"IFNG\", \"TNF\", \"IL2RG\", \"CXCR3\",\n",
    "        # Naive/Memory\n",
    "        \"TCF7\", \"LEF1\", \"CCR7\", \"SELL\", \"IL7R\",\n",
    "        # B cell\n",
    "        \"CD19\", \"CD79A\", \"CD79B\", \"MS4A1\", \"CD20\",\n",
    "        \"PAX5\", \"BLK\", \"BANK1\",\n",
    "        # NK cell\n",
    "        \"NCAM1\", \"KLRD1\", \"KLRB1\", \"GNLY\"\n",
    "    ],\n",
    "\n",
    "    \"Plasma cell\": [\n",
    "        \"MZB1\", \"SDC1\", \"CD38\", \"PRDM1\", \"XBP1\",\n",
    "        \"TNFRSF17\", \"SLAMF7\", \"JCHAIN\", \"IRF4\",\n",
    "        \"IGHA1\", \"IGHA2\", \"IGHG1\", \"IGHG2\", \"IGHG3\", \"IGHG4\",\n",
    "        \"SSR4\", \"DERL3\"\n",
    "    ],\n",
    "\n",
    "    \"Macrophage/Histiocyte\": [\n",
    "        # Pan-macrophage\n",
    "        \"CD68\", \"CD163\", \"CSF1R\", \"AIF1\", \"ITGAM\",\n",
    "        # M2 macrophage\n",
    "        \"MRC1\", \"MSR1\", \"CD14\", \"FCGR3A\",\n",
    "        # Activation\n",
    "        \"C1QA\", \"C1QB\", \"C1QC\", \"APOE\", \"SPP1\",\n",
    "        # Tissue resident\n",
    "        \"CX3CR1\", \"LST1\", \"TYROBP\", \"FCER1G\",\n",
    "        \"CD74\", \"HLA-DRA\", \"HLA-DRB1\"\n",
    "    ],\n",
    "\n",
    "    \"Neutrophil\": [\n",
    "        \"S100A8\", \"S100A9\", \"S100A12\", \"LYZ\",\n",
    "        \"CEACAM8\", \"MPO\", \"ELANE\", \"CTSG\",\n",
    "        \"CSF3R\", \"FCGR3B\", \"CXCR2\", \"CXCR1\",\n",
    "        \"MMP9\", \"CAMP\"\n",
    "    ],\n",
    "\n",
    "    \"Adipocyte\": [\n",
    "        \"ADIPOQ\", \"LEP\", \"LPL\", \"PPARG\", \"CEBPA\",\n",
    "        \"FABP4\", \"PLIN1\", \"PLIN2\", \"CIDEC\",\n",
    "        \"FASN\", \"SCD\", \"RETN\", \"RBP4\"\n",
    "    ],\n",
    "\n",
    "    \"Other/Unknown\": []\n",
    "}\n",
    "class_list = {\n",
    "    0: \"epithelial\",\n",
    "    1: \"Basal/Myoepithelial\",\n",
    "    2: \"Smooth muscle\",\n",
    "    3: \"Fibroblast\",\n",
    "    4: \"Endothelial\",\n",
    "    5: \"Lymphocyte\",                # T + B ÌÜµÌï©\n",
    "    6: \"Plasma cell\",\n",
    "    7: \"Macrophage/Histiocyte\",     # ÌÜµÌï©\n",
    "    8: \"Neutrophil\",\n",
    "    9: \"Adipocyte\",\n",
    "    10: \"Other/Unknown\"\n",
    "}\n",
    "class_colors_hex = {\n",
    "    \"epithelial\": \"#FF0000\",        # Îπ®Í∞ï\n",
    "    \"Basal/Myoepithelial\": \"#FFA500\",     # Ï£ºÌô©\n",
    "    \"Smooth muscle\": \"#8B4513\",           # Í∞àÏÉâ\n",
    "    \"Fibroblast\": \"#00FF00\",              # Ï¥àÎ°ù\n",
    "    \"Endothelial\": \"#0000FF\",             # ÌååÎûë\n",
    "    \"Lymphocyte\": \"#FFFF00\",              # ÎÖ∏Îûë (T/B lymphocyte ÌÜµÌï©)\n",
    "    \"Plasma cell\": \"#9400D3\",             # Î≥¥Îùº\n",
    "    \"Macrophage/Histiocyte\": \"#00FFFF\",   # ÏãúÏïà(Ï≤≠Î°ù)\n",
    "    \"Neutrophil\": \"#1E90FF\",              # DodgerBlue (Î∞ùÏùÄ ÌååÎûë)\n",
    "    \"Adipocyte\": \"#FFC0CB\",               # ÌïëÌÅ¨\n",
    "    \"Other/Unknown\": \"#808080\"            # ÌöåÏÉâ\n",
    "}\n",
    "\n",
    "class_colors = {\n",
    "    \"epithelial\": [255, 0, 0],            # Îπ®Í∞ï - Ï¢ÖÏñë ÏÉÅÌîº\n",
    "    \"Basal/Myoepithelial\": [255, 165, 0],       # Ï£ºÌô©\n",
    "    \"Smooth muscle\": [139, 69, 19],             # Í∞àÏÉâ\n",
    "    \"Fibroblast\": [0, 255, 0],                  # Ï¥àÎ°ù\n",
    "    \"Endothelial\": [0, 0, 255],                 # ÌååÎûë\n",
    "    \"Lymphocyte\": [255, 255, 0],                # ÎÖ∏Îûë (T/B ÌÜµÌï©)\n",
    "    \"Plasma cell\": [148, 0, 211],               # Î≥¥Îùº\n",
    "    \"Macrophage/Histiocyte\": [0, 255, 255],     # ÏãúÏïà (Ï≤≠Î°ù)\n",
    "    \"Neutrophil\": [30, 144, 255],               # ÎèÑÏ†ÄÎ∏îÎ£® (Î∞ùÏùÄ ÌååÎûë)\n",
    "    \"Adipocyte\": [255, 192, 203],               # ÌïëÌÅ¨\n",
    "    \"Other/Unknown\": [128, 128, 128]            # ÌöåÏÉâ\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f8c3b9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TENX ÎßàÏª§ gene list\n",
    "marker_genes = {\n",
    "    \"epithelial\": [\n",
    "        # Epithelial Í∏∞Î≥∏ ÎßàÏª§\n",
    "        \"EPCAM\", \"KRT8\", \"KRT18\", \"KRT19\", \"KRT7\",\n",
    "        \"TACSTD2\", \"CLDN4\",\n",
    "        # Luminal ÎßàÏª§\n",
    "        \"ERBB2\", \"ESR1\", \"PGR\", \"GATA3\",\n",
    "        # Proliferation\n",
    "        \"MKI67\", \"PCNA\",\n",
    "        # Junction\n",
    "        \"CDH1\",\n",
    "        # Keratin\n",
    "        \"KRT5\", \"KRT14\", \"KRT15\", \"KRT16\", \"KRT17\", \"KRT20\", \"KRT23\", \"KRT80\",\n",
    "        # Ï∂îÍ∞Ä\n",
    "        \"MSLN\", \"WFDC2\"\n",
    "    ],\n",
    "\n",
    "    \"Basal/Myoepithelial\": [\n",
    "        # Basal specific\n",
    "        \"KRT5\", \"KRT14\", \"KRT15\", \"KRT17\", \"TP63\",\n",
    "        # Myoepithelial specific\n",
    "        \"ACTA2\", \"TAGLN\", \"MYH11\", \"CNN1\",\n",
    "        # Ï∂îÍ∞Ä\n",
    "        \"PDGFRB\", \"RGS5\", \"COL17A1\", \"ITGA6\", \"LAMC2\"\n",
    "    ],\n",
    "\n",
    "    \"Fibroblast\": [\n",
    "        # ECM Í¥ÄÎ†®\n",
    "        \"COL1A1\", \"COL1A2\", \"COL3A1\", \"COL5A2\", \n",
    "        \"COL6A1\", \"COL6A2\", \"COL11A1\",\n",
    "        \"DCN\", \"LUM\", \"BGN\", \"SPARC\", \"FN1\",\n",
    "        # Fibroblast specific\n",
    "        \"PDGFRA\", \"FAP\", \"CXCL12\", \"DPT\",\n",
    "        # CAF ÎßàÏª§\n",
    "        \"MMP2\", \"MMP14\",\n",
    "        # Ï∂îÍ∞Ä\n",
    "        \"C7\", \"C1R\", \"C1S\", \"FBLN1\", \"POSTN\", \"THY1\"\n",
    "    ],\n",
    "\n",
    "    \"Endothelial\": [\n",
    "        # Pan-endothelial\n",
    "        \"PECAM1\", \"VWF\", \"CDH5\", \"CD34\", \"CLDN5\",\n",
    "        # Vascular\n",
    "        \"KDR\", \"FLT1\", \"ENG\", \"EGFL7\",\n",
    "        # Capillary/Lymphatic\n",
    "        \"PLVAP\", \"ACKR1\", \"PROX1\", \"LYVE1\",\n",
    "        # Ï∂îÍ∞Ä\n",
    "        \"RAMP2\", \"SELE\", \"ECSCR\"\n",
    "    ],\n",
    "\n",
    "    \"Lymphocyte\": [\n",
    "        # T cell Î≤îÏö©\n",
    "        \"PTPRC\", \"CD3D\", \"CD3E\", \"CD3G\", \n",
    "        \"TRAC\", \"TRBC1\", \"TRBC2\",\n",
    "        # T cell markers\n",
    "        \"CD4\", \"CD8A\", \"CD8B\",\n",
    "        \"IL7R\", \"CCR7\", \"LTB\",\n",
    "        # Cytotoxic/NK\n",
    "        \"NKG7\", \"GNLY\", \"PRF1\", \"GZMK\", \"GZMB\", \"GZMH\", \"KLRD1\",\n",
    "        # B cell\n",
    "        \"MS4A1\", \"CD79A\", \"CD79B\", \"CD19\", \"CD74\",\n",
    "        # Ï∂îÍ∞Ä\n",
    "        \"CD2\", \"CD5\", \"CD7\", \"CD27\", \"CD28\", \"CD52\", \"CD69\"\n",
    "    ],\n",
    "\n",
    "    \"Plasma cell\": [\n",
    "        \"MZB1\", \"XBP1\", \"JCHAIN\",\n",
    "        \"SDC1\", \"TNFRSF17\",\n",
    "        # Immunoglobulin\n",
    "        \"IGKC\", \"IGLC3\",\n",
    "        \"IGHG1\", \"IGHG3\", \"IGHG4\", \"IGHM\",\n",
    "        # Ï∂îÍ∞Ä\n",
    "        \"DERL3\", \"SEC11C\", \"SSR4\"\n",
    "    ],\n",
    "\n",
    "    \"Macrophage/Histiocyte\": [\n",
    "        # Pan-macrophage\n",
    "        \"LYZ\", \"CD68\", \"CSF1R\", \"SPI1\", \"AIF1\",\n",
    "        # M2/Tissue resident\n",
    "        \"CD163\", \"MRC1\", \"MS4A7\", \"FCGR3A\",\n",
    "        # Complement\n",
    "        \"C1QA\", \"C1QB\", \"C1QC\",\n",
    "        # Ï∂îÍ∞Ä\n",
    "        \"TYROBP\", \"FCER1G\", \"CTSS\", \"LST1\",\n",
    "        \"MARCO\", \"APOE\", \"TREM2\", \"IL1B\", \"CD14\"\n",
    "    ],\n",
    "\n",
    "    \"Neutrophil\": [\n",
    "        \"S100A8\", \"S100A9\", \"S100A12\",\n",
    "        \"CXCR2\", \"CXCR1\",\n",
    "        \"CTSG\", \"LTF\", \"LCN2\",\n",
    "        \"MNDA\", \"CSF3R\"\n",
    "    ],\n",
    "\n",
    "    \"Adipocyte\": [\n",
    "        \"ADIPOQ\", \"FABP4\", \"LPL\", \"PPARG\", \"RBP4\",\n",
    "        \"LEP\", \"PLIN4\"\n",
    "    ],\n",
    "\n",
    "    \"Other/Unknown\": []\n",
    "}\n",
    "class_list = {\n",
    "    0: \"epithelial\",\n",
    "    1: \"Basal/Myoepithelial\",\n",
    "    2: \"Smooth muscle\",\n",
    "    3: \"Fibroblast\",\n",
    "    4: \"Endothelial\",\n",
    "    5: \"Lymphocyte\",                # T + B ÌÜµÌï©\n",
    "    6: \"Plasma cell\",\n",
    "    7: \"Macrophage/Histiocyte\",     # ÌÜµÌï©\n",
    "    8: \"Neutrophil\",\n",
    "    9: \"Adipocyte\",\n",
    "    10: \"Other/Unknown\"\n",
    "}\n",
    "class_colors_hex = {\n",
    "    \"epithelial\": \"#FF0000\",        # Îπ®Í∞ï\n",
    "    \"Basal/Myoepithelial\": \"#FFA500\",     # Ï£ºÌô©\n",
    "    \"Smooth muscle\": \"#8B4513\",           # Í∞àÏÉâ\n",
    "    \"Fibroblast\": \"#00FF00\",              # Ï¥àÎ°ù\n",
    "    \"Endothelial\": \"#0000FF\",             # ÌååÎûë\n",
    "    \"Lymphocyte\": \"#FFFF00\",              # ÎÖ∏Îûë (T/B lymphocyte ÌÜµÌï©)\n",
    "    \"Plasma cell\": \"#9400D3\",             # Î≥¥Îùº\n",
    "    \"Macrophage/Histiocyte\": \"#00FFFF\",   # ÏãúÏïà(Ï≤≠Î°ù)\n",
    "    \"Neutrophil\": \"#1E90FF\",              # DodgerBlue (Î∞ùÏùÄ ÌååÎûë)\n",
    "    \"Adipocyte\": \"#FFC0CB\",               # ÌïëÌÅ¨\n",
    "    \"Other/Unknown\": \"#808080\"            # ÌöåÏÉâ\n",
    "}\n",
    "\n",
    "class_colors = {\n",
    "    \"epithelial\": [255, 0, 0],            # Îπ®Í∞ï - Ï¢ÖÏñë ÏÉÅÌîº\n",
    "    \"Basal/Myoepithelial\": [255, 165, 0],       # Ï£ºÌô©\n",
    "    \"Smooth muscle\": [139, 69, 19],             # Í∞àÏÉâ\n",
    "    \"Fibroblast\": [0, 255, 0],                  # Ï¥àÎ°ù\n",
    "    \"Endothelial\": [0, 0, 255],                 # ÌååÎûë\n",
    "    \"Lymphocyte\": [255, 255, 0],                # ÎÖ∏Îûë (T/B ÌÜµÌï©)\n",
    "    \"Plasma cell\": [148, 0, 211],               # Î≥¥Îùº\n",
    "    \"Macrophage/Histiocyte\": [0, 255, 255],     # ÏãúÏïà (Ï≤≠Î°ù)\n",
    "    \"Neutrophil\": [30, 144, 255],               # ÎèÑÏ†ÄÎ∏îÎ£® (Î∞ùÏùÄ ÌååÎûë)\n",
    "    \"Adipocyte\": [255, 192, 203],               # ÌïëÌÅ¨\n",
    "    \"Other/Unknown\": [128, 128, 128]            # ÌöåÏÉâ\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb7b6b46",
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify_cell_by_genes(gene_list, marker_dict, min_markers=1, min_score_ratio=0.5):\n",
    "    \"\"\"\n",
    "    Ïó¨Îü¨ Ïú†Ï†ÑÏûêÎ•º Í∏∞Î∞òÏúºÎ°ú cell type scoring\n",
    "    \n",
    "    Args:\n",
    "        gene_list: ÏÖÄ ÎÇ¥ Í≤ÄÏ∂úÎêú Ïú†Ï†ÑÏûê Î¶¨Ïä§Ìä∏\n",
    "        marker_dict: cell typeÎ≥Ñ marker gene ÎîïÏÖîÎÑàÎ¶¨\n",
    "        min_markers: cell typeÏúºÎ°ú Î∂ÑÎ•òÌïòÍ∏∞ ÏúÑÌïú ÏµúÏÜå marker Ïàò\n",
    "        min_score_ratio: 2Îì±Í≥ºÏùò Ï†êÏàò ÎπÑÏú® (0.5 = 2Îì±Ïù¥ 1Îì±Ïùò 50% Ïù¥ÏÉÅÏù¥Î©¥ Ïï†Îß§Ìï®)\n",
    "    \"\"\"\n",
    "    scores = {cell_type: 0 for cell_type in marker_dict.keys()}\n",
    "    \n",
    "    # Í∞Å geneÏù¥ Ïñ¥Îäê cell typeÏùò markerÏù∏ÏßÄ Ïπ¥Ïö¥Ìä∏\n",
    "    for gene in gene_list:\n",
    "        for cell_type, markers in marker_dict.items():\n",
    "            if gene in markers:\n",
    "                scores[cell_type] += 1\n",
    "    \n",
    "    # Ï†êÏàò Ï†ïÎ†¨\n",
    "    sorted_scores = sorted(scores.items(), key=lambda x: x[1], reverse=True)\n",
    "    max_score = sorted_scores[0][1]\n",
    "    second_score = sorted_scores[1][1] if len(sorted_scores) > 1 else 0\n",
    "    \n",
    "    # ÏµúÏÜå marker Ïàò ÎØ∏ÎßåÏù¥Î©¥ Unknown\n",
    "    if max_score < min_markers:\n",
    "        return 'Other/Unknown', scores\n",
    "    \n",
    "    # 2Îì±Í≥º Ï†êÏàò Ï∞®Ïù¥Í∞Ä ÏûëÏúºÎ©¥ Ïï†Îß§Ìïú Í≤ΩÏö∞\n",
    "    if second_score > 0 and second_score / max_score >= min_score_ratio:\n",
    "        # ÎèôÏ†êÏóê Í∞ÄÍπåÏö¥ Í≤ΩÏö∞ Ïö∞ÏÑ†ÏàúÏúÑ Ï†ÅÏö©\n",
    "        tied_types = [ct for ct, score in scores.items() \n",
    "                     if score >= max_score * min_score_ratio]\n",
    "        \n",
    "        priority = [\n",
    "            \"epithelial\", \"Cycling\", \"Lymphocyte\", \"Plasma cell\",\n",
    "            \"Macrophage/Histiocyte\", \"Neutrophil\", \"CAF_like\",\n",
    "            \"Fibroblast\", \"Pericyte/SMC\", \"Endothelial\",\n",
    "            \"Basal/Myoepithelial\", \"Smooth muscle\", \"Adipocyte\"\n",
    "        ]\n",
    "        \n",
    "        for cell_type in priority:\n",
    "            if cell_type in tied_types:\n",
    "                return cell_type, scores\n",
    "    \n",
    "    # Î™ÖÌôïÌïú 1Îì±\n",
    "    best_type = max(scores, key=scores.get)\n",
    "    return best_type, scores\n",
    "\n",
    "xenium_annotation_list = glob('../../data/spatialTranscriptome/cellvit_seg/*_cellvit_seg.parquet')\n",
    "save_path = '../../data/spatialTranscriptome/preprocessed_xenium/'\n",
    "xenium_transcripts_list = [f.replace(\"cellvit_seg/\", \"transcripts/\") for f in xenium_annotation_list]\n",
    "xenium_transcripts_list = [f.replace(\"_cellvit_seg.parquet\", \"_transcripts.parquet\") for f in xenium_transcripts_list]\n",
    "\n",
    "for i in range(len(xenium_transcripts_list)):\n",
    "    if os.path.exists(save_path + 'labels/' + os.path.basename(xenium_annotation_list[i]).replace('_cellvit_seg.parquet', '.csv')):\n",
    "        print(f\"Skipping (already processed): {save_path + 'labels/' + os.path.basename(xenium_annotation_list[i]).replace('_cellvit_seg.parquet', '.csv')}\")\n",
    "        continue\n",
    "    \n",
    "    xenium_transcript_path = xenium_transcripts_list[i]\n",
    "    df_transcript = pd.read_parquet(xenium_transcript_path)\n",
    "    df_filtered = df_transcript[df_transcript['qv'] > QV_THRESHOLD].copy()\n",
    "    \n",
    "    # feature_name Ï≤òÎ¶¨\n",
    "    if type(df_filtered['feature_name'].iloc[0]) == bytes:\n",
    "        df_filtered['feature_name'] = df_filtered['feature_name'].str.decode('utf-8')\n",
    "        df_filtered = df_filtered[~df_filtered['feature_name'].str.contains('BLANK|NegControl|antisense', case=False, na=False)]\n",
    "    elif type(df_filtered['feature_name'].iloc[0]) == str:\n",
    "        df_filtered = df_filtered[~df_filtered['feature_name'].str.contains('BLANK|NegControl|antisense', case=False, na=False)]\n",
    "    else:\n",
    "        print(f\"Skipping: {xenium_transcript_path}\")\n",
    "        continue\n",
    "    \n",
    "    # Segmentation Îç∞Ïù¥ÌÑ∞ Î°úÎìú\n",
    "    xenium_annotation_path = xenium_annotation_list[i]\n",
    "    df_seg = pd.read_parquet(xenium_annotation_path)\n",
    "    \n",
    "    # GeoDataFrame Î≥ÄÌôò\n",
    "    transcript_gdf = gpd.GeoDataFrame(\n",
    "        df_filtered,\n",
    "        geometry=gpd.points_from_xy(df_filtered['he_x'], df_filtered['he_y'])\n",
    "    )\n",
    "    \n",
    "    polygons = [wkb.loads(geom) for geom in df_seg['geometry']]\n",
    "    seg_gdf = gpd.GeoDataFrame(df_seg, geometry=polygons)\n",
    "    \n",
    "    # Spatial join ÏàòÌñâ\n",
    "    joined = gpd.sjoin(transcript_gdf, seg_gdf, how='inner', predicate='within')\n",
    "    \n",
    "    # üöÄ ÏµúÏ†ÅÌôî: groupbyÎ°ú Ìïú Î≤àÏóê Ï≤òÎ¶¨\n",
    "    annotations = []\n",
    "    \n",
    "    if len(joined) > 0:\n",
    "        # index_rightÎ°ú Í∑∏Î£πÌôîÌïòÏó¨ Í∞Å ÏÖÄÏùò Ïú†Ï†ÑÏûê Î¶¨Ïä§Ìä∏ Ìïú Î≤àÏóê ÏÉùÏÑ±\n",
    "        grouped = joined.groupby('index_right')['feature_name'].apply(list)\n",
    "        \n",
    "        for idx in tqdm(range(len(seg_gdf))):\n",
    "            if idx in grouped.index:\n",
    "                genes_in_cell = grouped[idx]\n",
    "                cell_type, score = classify_cell_by_genes(genes_in_cell, marker_genes)\n",
    "                \n",
    "                bounds = seg_gdf.iloc[idx].geometry.bounds\n",
    "                annotations.append({\n",
    "                    'x1': int(bounds[0]),\n",
    "                    'y1': int(bounds[1]),\n",
    "                    'x2': int(bounds[2]),\n",
    "                    'y2': int(bounds[3]),\n",
    "                    'class_name': cell_type,\n",
    "                    'gene_count': len(genes_in_cell)\n",
    "                })\n",
    "    \n",
    "    # Ï†ÄÏû•\n",
    "    if annotations:\n",
    "        df = pd.DataFrame(annotations)\n",
    "        \n",
    "        # ÌÜµÍ≥Ñ Ï∂úÎ†•\n",
    "        print(f\"\\n=== {os.path.basename(xenium_annotation_path)} ===\")\n",
    "        print(f\"Total cells: {len(df)}\")\n",
    "        print(f\"\\nCell type distribution:\")\n",
    "        type_counts = df['class_name'].value_counts()\n",
    "        for cell_type, count in type_counts.items():\n",
    "            percentage = (count / len(df)) * 100\n",
    "            print(f\"  {cell_type}: {count} ({percentage:.1f}%)\")\n",
    "        \n",
    "        # Unknown ÎπÑÏú®Ïù¥ ÎÜíÏúºÎ©¥ Í≤ΩÍ≥†\n",
    "        unknown_ratio = type_counts.get('Other/Unknown', 0) / len(df)\n",
    "        if unknown_ratio > 0.5:\n",
    "            print(f\"\\n‚ö†Ô∏è WARNING: {unknown_ratio*100:.1f}% cells are Unknown!\")\n",
    "            print(\"Consider:\")\n",
    "            print(\"  1. Lowering min_markers threshold\")\n",
    "            print(\"  2. Adding more marker genes\")\n",
    "            print(\"  3. Checking transcript quality\")\n",
    "        \n",
    "        # Ï†ÄÏû•\n",
    "        create_dir(save_path + 'labels/')\n",
    "        output_file = save_path + 'labels/' + os.path.basename(xenium_annotation_path).replace('_cellvit_seg.parquet', '.csv')\n",
    "        df.to_csv(output_file, index=False)\n",
    "\n",
    "print(\"All samples processed!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7697cf9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import openslide as ops\n",
    "\n",
    "reduction_factor=20\n",
    "slide=ops.OpenSlide('../../data/spatialTranscriptome/preprocessed_xenium/wsis/TENX106.tif')\n",
    "thumbnail = slide.get_thumbnail((slide.level_dimensions[0][0] // reduction_factor, slide.level_dimensions[0][1] // reduction_factor))\n",
    "mask=np.ones_like(np.array(thumbnail)) * 0\n",
    "labels_df=pd.read_csv('../../data/spatialTranscriptome/preprocessed_xenium/labels/TENX106.csv')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(22, 20))\n",
    "\n",
    "for idx, row in labels_df.iterrows():\n",
    "    x=row['x1']/reduction_factor + row['x2']/reduction_factor\n",
    "    x=x//2\n",
    "    y=row['y1']/reduction_factor + row['y2']/reduction_factor\n",
    "    y=y//2\n",
    "    \n",
    "    mask[int(y):int(y)+2, int(x):int(x)+2]=np.array([class_colors[row['class_name']]])/255.\n",
    "\n",
    "ax.imshow(mask*0.5 + np.array(thumbnail)/255.*0.5)\n",
    "ax.axis('off')\n",
    "ax.set_title('Cell Type Annotation', fontsize=16, fontweight='bold')\n",
    "\n",
    "# ÌÅ¥ÎûòÏä§Î≥Ñ Í∞úÏàò Í≥ÑÏÇ∞\n",
    "class_counts = labels_df['class_name'].value_counts()\n",
    "\n",
    "# Î≤îÎ°Ä Ï∂îÍ∞Ä (ÌÅ¥ÎûòÏä§ Í∞úÏàò Ìè¨Ìï®)\n",
    "legend_patches = []\n",
    "for class_name, hex_color in class_colors_hex.items():\n",
    "    count = class_counts.get(class_name, 0)\n",
    "    label = f\"{class_name}: {count}\"\n",
    "    patch = mpatches.Patch(color=hex_color, label=label)\n",
    "    legend_patches.append(patch)\n",
    "\n",
    "ax.legend(handles=legend_patches, \n",
    "         loc='upper right', \n",
    "         fontsize=15,\n",
    "         framealpha=0.95,\n",
    "         bbox_to_anchor=(1.18, 1.0),\n",
    "         title='Cell Type (Count)',\n",
    "         title_fontsize=12)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "# Ï†ÑÏ≤¥ ÌÜµÍ≥Ñ Ï∂úÎ†•\n",
    "print(\"=== Cell Type Statistics ===\")\n",
    "print(f\"Total cells: {len(labels_df)}\")\n",
    "print(\"\\nCell type distribution:\")\n",
    "print(class_counts.sort_index())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21a3d4ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_filtered[df_filtered['cell_id']==b'aaaeppaj-1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f7a26e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_transcripts"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
